src/traits/gomoku_ai.rs

```diff
@@ -0,0 +1,78 @@
+use std::{collections::HashMap, path::Path};
+
+use crate::model::{GameType, Stone};
+
+/// Trait which defines possible interactions with a Gomoku AI
+pub trait GomokuAI<CellType, IndexType = usize> {
+    /// Start the AI with a given square board size. The AI
+    /// is allowed to refuse playing on with this board
+    /// size, in which case it should return `None`.
+    fn start(board_size: IndexType) -> Option<Self>
+    where
+        Self: Sized;
```

Comment:
Where does the `Sized` in `Self: Sized;` come from ?

Reply:
`Sized` is a trait defined in standard Rust, you can check its docs [here](https://doc.rust-lang.org/std/marker/trait.Sized.html).

It only imposes that the object has a constant and predictable size in local memory, which is true to all concrete types, including `Vec<T>`  for example, which does manage dynamic memory, but locally only contains a pointer and other metadata of predictable size.
The only case where a type doesn't match this trait is for [trait objects](https://doc.rust-lang.org/book/ch17-02-trait-objects.html), (dereferenced references to any object implementing a trait)

Since `Option<T>` stores the optional object in place, it needs to have a fixed size, so `Option` requires that its type parameters implements `Sized`.
By writing `where Self: Sized`, we explicitly propagate the constraint, which we must do because Rust does not automatically propagate constraints, by design.

Reply:
Very well